### ISSUE TITLE
Fine-Tuning with Pretrained Model in Docker Image

### ISSUE THREAD

**Issue by JianYang93** [zhang0jhon/AttentionOCR#70]
Hi @zhang0jhon 

Thanks for the great work! The pretrained model you put in the docker image worked pretty well!
However I want to fine-tune the parameters for my specific application.
Is it possible to fine-tune the pretrained model? Since the model is different between the docker image and here, do you have the traning code for the docker version and/or do you have a fine-tuning code already?

Thanks in advance!

**Comment by JianYang93** [comment]
Seems I figured out a way to do it. 

The pretrained model is in a frozen graph file (.pb) and the parameter values are in type tf.constant. One can set them to variables and try to train them, but in this case, the frozen-graph format has lost information regarding the while-context and gradients cannot be calculated. (I have tried and failed. The reason is in https://github.com/tensorflow/tensorflow/issues/7404) 

However, the author has provided the inception model structure code for the docker version in the docker image, so theoretically one can reconstruct the graph and then import the parameter values from the pretrained model. Specifically, I did the following:
(1) constructed a new graph based on the inception model code provided by the author in docker
(2) loaded the pretrained model (.pb file) to another graph and got all the model parameter values
(3) assigned these values to the variables in the first graph

I tested that the outputs given by the pretrained model and the new graph I constructed were the same. Also the new model can be trained with additional inputs, although I have not tested whether accuracy can be improved.

**Comment by trungpham2606** [comment]
@JianYang93 
Can you please send me the pretrained weights ?

**Comment by JianYang93** [comment]
@trungpham2606 That's a pretty big file and I am unable to upload here! You can simply load the graph from the pb file the author shared in docker, and then read out all the constants in the graph.

**Comment by trungpham2606** [comment]
@JianYang93 
Unfortunately that I dont use Docker :(
Can you please send me the weights via my gmail: trungpham2606@gmail.com
Thank you so much !

**Comment by JianYang93** [comment]
@trungpham2606 Sure. Please check you email

**Comment by johnbird7** [comment]
@JianYang93 i have to docker weights text_recognition_5435.pb but i dont know how to finetune on my images. would you please guide me? 

a. how do i create a dataset to finetune, where do i specify ground truth, split for train and validation?
b. which script to run after creating own dataset?

plz help..

**Comment by JianYang93** [comment]
Hi,

You may want to look at the dataset.py and see how the author organized the labels. In dataset.py the author constructed four classes for ArT, ReCTS, LSVT and ICDAR2017 respectively. Basically you will need the sequence labels and polygon coordinates and then you can get the mask and bbox coordinates from them. You can download one of the datasets and run the dataset.py and look at the variables to get a feeling. After that you can mimic those four classes and write your own one.

Loading the dataset involves the functions get_roidb and get_batch_train_dataflow in text_dataflow.py. I guess you can do the train_val split somewhere there. 

Hope it hels!

**Comment by johnbird7** [comment]
@JianYang93 thank you for the info on creation of dataset, it helped. Would you please send me the ckpt and new pb that you have created which is traineable/finetunble on own dataset? my email id is johnbird77777@gmail.com

awaiting your response, Thank you. have a nice day :)

**Comment by JianYang93** [comment]
@johnbird7 

No problem, glad it helps!

Sorry I do not have the ckpt file on hand. I can share how I did it

Sine you have the pb file, I assume you have the codes in the docker image. I used the inception_model function in model.inception_v4_model_padding_same.py to build the computation graph. Then I get all the variables names in the graph by  tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES)

After that I loaded the pretrained model in another graph, ran all the corresponding tensors to get the numpy array values for all the variables. 

Finally I went back to the first graph I built and ran assignment operations for all those variables. To make sure things are consistent I used a random image and looked at the output features for the new graph and the pretrained model and compare them.

**Comment by johnbird7** [comment]
@JianYang93  i'm sorry, im not very good at tensorflow. could you possibly generate the ckpt for me please. sorry if im asking too much but i really need to finetune with my own data asap. :(

**Comment by JianYang93** [comment]
@johnbird7 Sorry, for some personal reasons I am not supposed to send out things I did. I can share the original pretrained model from the author but I guess you already have it. 

Best

**Comment by johnbird7** [comment]
@JianYang93 i understand. it's ok i'll train the model with the huge datasets from scratch.. ill use the default params he provided in train.py script until that checkpoint is reached. ill add in my own data to that huge dataset. if i run into any errrors plz help.. thank you, have a good time.

**Comment by JianYang93** [comment]
@johnbird7 I will if I can... I never tried training with the code the author provided. However I once tried training a model built on top of this recognition model with all the data (20G or more..) but had no luck. Please also let us know if you get any insights in how to train the model.

**Comment by whereitogo** [comment]
> Seems I figured out a way to do it.
> 
> The pretrained model is in a frozen graph file (.pb) and the parameter values are in type tf.constant. One can set them to variables and try to train them, but in this case, the frozen-graph format has lost information regarding the while-context and gradients cannot be calculated. (I have tried and failed. The reason is in [tensorflow/tensorflow#7404](https://github.com/tensorflow/tensorflow/issues/7404))
> 
> However, the author has provided the inception model structure code for the docker version in the docker image, so theoretically one can reconstruct the graph and then import the parameter values from the pretrained model. Specifically, I did the following:
> (1) constructed a new graph based on the inception model code provided by the author in docker
> (2) loaded the pretrained model (.pb file) to another graph and got all the model parameter values
> (3) assigned these values to the variables in the first graph
> 
> I tested that the outputs given by the pretrained model and the new graph I constructed were the same. Also the new model can be trained with additional inputs, although I have not tested whether accuracy can be improved.

hi,I try to train on my datasets but its slowly(30min one epoch).Can you please send me the weights of AttentionOCR? Thank you so much !my email:15765058073fu@gmail.com

**Comment by naraklog** [comment]
@JianYang93
I need the saved model for a project and I am not very familiar with docker. Would you please send it to my Gmail? karansingh8777@gmail.com
Awaiting your response, Thank you
